### ISSUE TITLE
[torch.disitributed] Persistent Deadlock when overlapping NCCL P2P and Collectives from separate Python Threads

### ISSUE THREAD

**Issue by ilmarkov** [pytorch/pytorch#175145]
### 🐛 Describe the bug

Launching NCCL Point-to-Point (P2P) operations (isend/irecv) and Collective operations (all_reduce) on separate Python threads results in a permanent deadlock, even when using separate ProcessGroup instances, dedicated CUDA streams, and explicit thread synchronization.

The issue persists despite limiting hardware resources (e.g., NCCL_MAX_NCHANNELS=1), suggesting a circular dependency within the NCCL internal Progress Engine (Proxy Thread).

Repro script.  Launch with `torchrun --nproc_per_node=4 nccl_threaded_p2p_vs_allreduce_repro.py --bg-p2p-ops 16 --tensor-numel 1024 [--use-fence]`

"""
#!/usr/bin/env python3

"""
Shape:
- Main thread/stream: all_reduce on pg_main.
- Background thread/stream: send/recv pairs on pg_bg.
- Separate process groups, one per thread role.

Notes:
- `--use-fence` enables event fencing and host-side sync on BG thread.

Example:
  torchrun --nproc-per-node=4 \
    nccl_threaded_p2p_vs_allreduce_repro.py \
    --max-iters 10000 --log-every 100 --bg-p2p-ops 16
"""

import argparse
import os
import threading
import traceback
from typing import Any

import torch
import torch.distributed as dist

def _warmup(
    *,
    pg_main: dist.ProcessGroup,
    pg_bg: dist.ProcessGroup,
    rank: int,
    world: int,
    local_rank: int,
) -> None:
    warm = torch.ones((8,), device="cuda", dtype=torch.float32)
    dist.all_reduce(warm, group=pg_main)
    dist.all_reduce(warm, group=pg_bg)

    # Warmup one P2P pair on pg_bg.
    if world % 2 == 0:
        peer = rank ^ 1
    else:
        peer = (rank + 1) % world
    send = torch.full((8,), float(rank), device="cuda", dtype=torch.float32)
    recv = torch.empty_like(send)
    if rank < peer:
        ops = [
            dist.P2POp(dist.isend, send, peer, group=pg_bg),
            dist.P2POp(dist.irecv, recv, peer, group=pg_bg),
        ]
    else:
        ops = [
            dist.P2POp(dist.irecv, recv, peer, group=pg_bg),
            dist.P2POp(dist.isend, send, peer, group=pg_bg),
        ]
    reqs = dist.batch_isend_irecv(ops)
    for req in reqs:
        req.wait()

    torch.cuda.synchronize()
    dist.barrier(device_ids=[local_rank])

def _main_loop(
    *,
    stream: torch.cuda.Stream,
    pg_main: dist.ProcessGroup,
    local_rank: int,
    rank: int,
    world: int,
    tensor_numel: int,
    start_gate: threading.Barrier,
    stop_event: threading.Event,
    max_iters: int,
    log_every: int,
    use_fence: bool,
    fence_state: dict[str, Any],
    fence_lock: threading.Lock,
) -> None:
    torch.cuda.set_device(local_rank)
    x = torch.full((tensor_numel,), float(rank + 1), device="cuda", dtype=torch.float32)
    start_gate.wait()
    print(f"rank={rank} thread=main started", flush=True)

    i = 0
    try:
        while not stop_event.is_set() and i < max_iters:
            with torch.cuda.stream(stream):
                if use_fence:
                    with fence_lock:
                        bg_event = fence_state.get("bg_event")
                    if bg_event is not None:
                        stream.wait_event(bg_event)

                dist.all_reduce(x, group=pg_main)

                if use_fence:
                    ev = torch.cuda.Event(blocking=False)
                    ev.record(stream)
                    with fence_lock:
                        fence_state["main_event"] = ev

            if i % log_every == 0:
                print(f"rank={rank} thread=main iter={i}", flush=True)
            i += 1
    except Exception:
        print(f"rank={rank} thread=main exception", flush=True)
        traceback.print_exc()
        stop_event.set()

def _bg_loop(
    *,
    stream: torch.cuda.Stream,
    pg_bg: dist.ProcessGroup,
    local_rank: int,
    rank: int,
    world: int,
    tensor_numel: int,
    start_gate: threading.Barrier,
    stop_event: threading.Event,
    max_iters: int,
    log_every: int,
    bg_p2p_ops: int,
    use_fence: bool,
    fence_state: dict[str, Any],
    fence_lock: threading.Lock,
) -> None:
    torch.cuda.set_device(local_rank)

    if world % 2 == 0:
        peer = rank ^ 1
    else:
        peer = (rank + 1) % world
    send_first = rank < peer

    send_bufs = [
        torch.full((tensor_numel,), float(rank), device="cuda", dtype=torch.float32)
        for _ in range(bg_p2p_ops)
    ]
    recv_bufs = [torch.empty_like(send_bufs[0]) for _ in range(bg_p2p_ops)]

    start_gate.wait()
    print(
        f"rank={rank} thread=bg started peer={peer} bg_p2p_ops={bg_p2p_ops}",
        flush=True,
    )

    i = 0
    try:
        while not stop_event.is_set() and i < max_iters:
            with torch.cuda.stream(stream):
                if use_fence:
                    with fence_lock:
                        main_event = fence_state.get("main_event")
                    if main_event is not None:
                        stream.wait_event(main_event)

                if i == 0:
                    print(
                        f"rank={rank} thread=bg iter=0 launching_pairs={bg_p2p_ops}",
                        flush=True,
                    )

                for send_buf, recv_buf in zip(send_bufs, recv_bufs):
                    if send_first:
                        reqs = dist.batch_isend_irecv(
                            [dist.P2POp(dist.isend, send_buf, peer, group=pg_bg)]
                        )
                        for req in reqs:
                            req.wait()
                        reqs = dist.batch_isend_irecv(
                            [dist.P2POp(dist.irecv, recv_buf, peer, group=pg_bg)]
                        )
                    else:
                        reqs = dist.batch_isend_irecv(
                            [dist.P2POp(dist.irecv, recv_buf, peer, group=pg_bg)]
                        )
                        for req in reqs:
                            req.wait()
                        reqs = dist.batch_isend_irecv(
                            [dist.P2POp(dist.isend, send_buf, peer, group=pg_bg)]
                        )
                    for req in reqs:
                        req.wait()

                if use_fence:
                    ev = torch.cuda.Event(blocking=False)
                    ev.record(stream)
                    with fence_lock:
                        fence_state["bg_event"] = ev

            if use_fence:
                stream.synchronize()

            if i % log_every == 0:
                print(f"rank={rank} thread=bg iter={i}", flush=True)
            i += 1
    except Exception:
        print(f"rank={rank} thread=bg exception", flush=True)
        traceback.print_exc()
        stop_event.set()

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--tensor-numel", type=int, default=4096)
    parser.add_argument("--max-iters", type=int, default=100000)
    parser.add_argument("--log-every", type=int, default=100)
    parser.add_argument("--bg-p2p-ops", type=int, default=16)
    parser.add_argument("--use-fence", action="store_true")
    args = parser.parse_args()

    local_rank = int(os.environ.get("LOCAL_RANK", "0"))
    device_count = torch.cuda.device_count()
    if device_count <= 0:
        raise RuntimeError("No CUDA devices visible for NCCL repro.")
    if local_rank < 0 or local_rank >= device_count:
        raise RuntimeError(
            f"LOCAL_RANK={local_rank} is out of range for visible CUDA devices "
            f"(count={device_count})."
        )

    torch.cuda.set_device(local_rank)
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    world = dist.get_world_size()

    pg_main = dist.new_group(ranks=list(range(world)), backend="nccl")
    pg_bg = dist.new_group(ranks=list(range(world)), backend="nccl")

    print(
        f"rank={rank} local_rank={local_rank} world={world} "
        "starting minimal repro "
        f"bg_p2p_ops={args.bg_p2p_ops} use_fence={args.use_fence}",
        flush=True,
    )

    dist.barrier(device_ids=[local_rank])
    _warmup(
        pg_main=pg_main,
        pg_bg=pg_bg,
        rank=rank,
        world=world,
        local_rank=local_rank,
    )

    start_gate = threading.Barrier(2)
    stop_event = threading.Event()
    fence_state: dict[str, Any] = {"main_event": None, "bg_event": None}
    fence_lock = threading.Lock()

    main_stream = torch.cuda.Stream(priority=-1)
    bg_stream = torch.cuda.Stream(priority=0)

    bg_thread = threading.Thread(
        target=_bg_loop,
        kwargs=dict(
            stream=bg_stream,
            pg_bg=pg_bg,
            local_rank=local_rank,
            rank=rank,
            world=world,
            tensor_numel=args.tensor_numel,
            start_gate=start_gate,
            stop_event=stop_event,
            max_iters=args.max_iters,
            log_every=args.log_every,
            bg_p2p_ops=args.bg_p2p_ops,
            use_fence=args.use_fence,
            fence_state=fence_state,
            fence_lock=fence_lock,
        ),
        daemon=True,
    )
    bg_thread.start()

    try:
        _main_loop(
            stream=main_stream,
            pg_main=pg_main,
            local_rank=local_rank,
            rank=rank,
            world=world,
            tensor_numel=args.tensor_numel,
            start_gate=start_gate,
            stop_event=stop_event,
            max_iters=args.max_iters,
            log_every=args.log_every,
            use_fence=args.use_fence,
            fence_state=fence_state,
            fence_lock=fence_lock,
        )
    finally:
        stop_event.set()
        bg_thread.join()
        dist.barrier(device_ids=[local_rank])
        dist.destroy_process_group()

if __name__ == "__main__":
    main()
```

### Versions

Collecting environment information...
PyTorch version: 2.10.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 9.6 (Plow) (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: Could not collect
CMake version: version 4.2.1
Libc version: glibc-2.34

Python version: 3.12.9 (main, Aug 14 2025, 00:00:00) [GCC 11.5.0 20240719 (Red Hat 11.5.0-5)] (64-bit runtime)
Python platform: Linux-5.14.0-570.73.1.el9_6.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 590.48.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  160
On-line CPU(s) list:                     0-159
Vendor ID:                               GenuineIntel
Model name:                              Intel Xeon Processor (SapphireRapids)
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      40
Socket(s):                               2
Stepping:                                4
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Virtualization:                          VT-x
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               5 MiB (160 instances)
L1i cache:                               5 MiB (160 instances)
L2 cache:                                320 MiB (80 instances)
L3 cache:                                32 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-79
NUMA node1 CPU(s):                       80-159
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Unknown: No mitigations
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk

**Comment by ilmarkov** [comment]
The hypothesis is that in a single-process model, NCCL initializes a background CPU thread (Proxy Thread or ncclProxyService) which is shared for both threads regardless of mulitple processGroups.

When Thread A (Collective) and Thread B (P2P) both call into NCCL:
1. Thread A submits an all_reduce, and the Proxy begins coordinating the ring/tree across nodes.
2. Thread B submits an isend. The Proxy attempts to initiate the P2P connection handshake.
3. If Rank 0’s Proxy is busy waiting for a response from Rank 1 for the collective, but Rank 1’s Proxy is busy waiting for Rank 0 to acknowledge the P2P request, both Proxies stall.

`NCCL_MAX_NCHANNELS=1` or `NCCL_MAX_CTAS=1` does not resolve the issue, suggesting the bottleneck is not SM count but the CPU-side progress engine. `TORCH_NCCL_BLOCKING_WAIT=1` resolves the issue, which confirms the logic is sound but the engine cannot handle concurrent asynchronous submissions.

I wonder if there is a workaround in which it is possible to use torch.disitributed/NCCL from different threads for GPU-based inputs.

**Comment by ilmarkov** [comment]
The issue persists even if we use collectives (instead of point-to-point operations) in background thread.

**Comment by akhillanger** [comment]
As per official NCCL documentation, multi-thread use as in your example is not supported:
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/threadsafety.html

**Comment by sjeaugey** [comment]
@ilmarkov can you check whether CUMEM is disabled in your runs? I.e. is NCCL_CUMEM_ENABLE set to 0 in the environment? In doubt, you can run with NCCL_DEBUG=INFO and look for `NCCL_CUMEM_ENABLE` to see if NCCL reports it as being set to a non-default value.

**Comment by sjeaugey** [comment]
Also, regarding the problem statement:

>  using separate ProcessGroup instances, dedicated CUDA streams,

Using different CUDA streams does not guarantee independent progress for the different PGs. It only allows for concurrent progress.

> and explicit thread synchronization.

Can you comment on how this is implemented in the code above? This is the most critical part to ensure no deadlock, and I'm not sure I understand how it works in your example. Can you confirm what the different synchronization mechanisms are for, and how they ensure that the two threads post operations in a manner which cannot result in a non-deterministic submission order?
